server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/reservedstate/ReservedRepositoryAction.java

```diff
@@ -0,0 +1,112 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.client.internal.Requests.putRepositoryRequest;
+import static org.elasticsearch.common.xcontent.XContentHelper.mapToXContentParser;
+
+/**
+ * This Action is the reserved state save version of RestPutRepositoryAction/RestDeleteRepositoryAction
+ * <p>
+ * It is used by the ReservedClusterStateService to add/update or remove snapshot repositories. Typical usage
+ * for this action is in the context of file based settings.
+ */
+public class ReservedRepositoryAction implements ReservedClusterStateHandler<List<PutRepositoryRequest>> {
+    public static final String NAME = "snapshot_repositories";
+
+    private final RepositoriesService repositoriesService;
+
+    /**
+     * Creates a ReservedRepositoryAction
+     *
+     * @param repositoriesService requires RepositoriesService for the cluster state operations
+     */
+    public ReservedRepositoryAction(RepositoriesService repositoriesService) {
+        this.repositoriesService = repositoriesService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Collection<PutRepositoryRequest> prepare(Object input) {
+        List<PutRepositoryRequest> repositories = (List<PutRepositoryRequest>) input;
+
+        for (var repositoryRequest : repositories) {
+            validate(repositoryRequest);
+            RepositoriesService.validateRepositoryName(repositoryRequest.name());
+            repositoriesService.validateRepositoryCanBeCreated(repositoryRequest);
+        }
+
+        return repositories;
+    }
+
+    @Override
+    public TransformState transform(Object source, TransformState prevState) throws Exception {
+        var requests = prepare(source);
+
+        ClusterState state = prevState.state();
+
+        for (var request : requests) {
+            RepositoriesService.RegisterRepositoryTask task = new RepositoriesService.RegisterRepositoryTask(repositoriesService, request);
+            state = task.execute(state);
+        }
+
+        Set<String> entities = requests.stream().map(r -> r.name()).collect(Collectors.toSet());
+
+        Set<String> toDelete = new HashSet<>(prevState.keys());
+        toDelete.removeAll(entities);
+
+        for (var repositoryToDelete : toDelete) {
+            var task = new RepositoriesService.UnregisterRepositoryTask(repositoryToDelete);
+            state = task.execute(state);
+        }
+
+        return new TransformState(state, entities);
+
+    }
+
+    @Override
+    public List<PutRepositoryRequest> fromXContent(XContentParser parser) throws IOException {
+        List<PutRepositoryRequest> result = new ArrayList<>();
+
+        Map<String, ?> source = parser.map();
+
+        for (String name : source.keySet()) {
```

Comment:
Not too important but this is a bit of a strange loop, why not loop EntrySet and avoid the `source.get(name)`?